File: backend/app/schemas/agent.py
"""Agent 相关 Pydantic Schema。"""
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

# ---- 请求 ----

class AgentRunRequest(BaseModel):
    """Agent 运行请求。"""

    session_id: str = Field(default="default", max_length=64)
    question: str = Field(min_length=1, max_length=2000)


class AgentRunResumeRequest(BaseModel):
    """恢复运行请求，同时用于回答人工介入。

    run_id 指向之前中断的运行；decision 只在运行状态为 awaiting_review 时使用，
    accept 表示采纳当前草稿，rewrite 表示让 Generator 按反馈重写。
    """

    run_id: str = Field(description="之前中断或故障的 run_id")
    decision: str | None = Field(default=None, max_length=32)


# ---- 响应 ----

class AgentRunStepOut(BaseModel):
    """单个步骤的执行日志。"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    node_name: str
    attempt: int
    input_summary: str | None = None
    output_summary: str | None = None
    status: str  # running / success / failed
    error_message: str | None = None
    duration_ms: int | None = None
    started_at: datetime
    completed_at: datetime | None = None


class AgentRunOut(BaseModel):
    """Agent 运行记录响应。"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    run_id: str
    session_id: str
    question: str
    final_answer: str | None = None
    status: str  # pending / planning / retrieving / generating / reviewing / completed / failed
    error_message: str | None = None
    checkpoint_id: str | None = None
    retry_count: int
    started_at: datetime
    completed_at: datetime | None = None


class AgentRunDetailOut(AgentRunOut):
    """带步骤列表的运行详情。"""

    steps: list[AgentRunStepOut] = Field(default_factory=list)


class AgentRunStartResponse(BaseModel):
    """启动运行的响应（返回 run_id，前端用来轮询/连接 SSE）。

    status 为 awaiting_review 时 interrupt 非空，前端据此弹出人工裁决入口。
    """

    run_id: str
    status: str
    interrupt: dict[str, Any] | None = None


# ---- SSE 事件类型 ----

class AgentSSEEvent(BaseModel):
    """SSE 事件格式。"""

    event: str  # step_start / step_end / run_complete / run_failed
    data: dict

File: frontend/src/api/agent.ts
import api from "./index";

/** 人工介入的待裁决内容，由 Reviewer 质检不通过时产生。 */
export interface AgentInterrupt {
  type: string;
  question: string;
  draft: string;
  feedback: string;
  options: string[];
}

export interface AgentRunStartData {
  run_id: string;
  status: string;
  interrupt?: AgentInterrupt | null;
}

export interface AgentRunStep {
  node_name: string;
  attempt: number;
  input_summary: string | null;
  output_summary: string | null;
  status: string; // running / success / failed
  error_message: string | null;
  duration_ms: number | null;
  started_at: string | null;
  completed_at: string | null;
}

export interface AgentRunDetail {
  run: {
    run_id: string;
    session_id: string;
    question: string;
    final_answer: string | null;
    status: string;
    error_message: string | null;
    checkpoint_id: string | null;
    retry_count: number;
    started_at: string | null;
    completed_at: string | null;
  };
  steps: AgentRunStep[];
}

export interface AgentRunListItem {
  run_id: string;
  session_id: string;
  question: string;
  status: string;
  retry_count: number;
  started_at: string | null;
  completed_at: string | null;
}

export function startAgentRun(
  question: string,
  sessionId: string = "default"
): Promise<{ data: AgentRunStartData }> {
  return api.post("/agent/runs", { question, session_id: sessionId });
}

export function resumeAgentRun(
  runId: string,
  decision?: string
): Promise<{ data: AgentRunStartData }> {
  return api.post("/agent/runs/resume", { run_id: runId, decision });
}

export function fetchRunDetail(
  runId: string
): Promise<{ data: AgentRunDetail }> {
  return api.get(`/agent/runs/${runId}`);
}

export function listAgentRuns(
  sessionId: string = "default",
  limit: number = 20
): Promise<{ data: AgentRunListItem[] }> {
  return api.get("/agent/runs", { params: { session_id: sessionId, limit } });
}

// ---------------------------------------------------------------------------
// SSE 流式执行（fetch + ReadableStream）
// ---------------------------------------------------------------------------

export interface SSEEvent {
  type: string;
  run_id?: string;
  node?: string;
  output_summary?: string;
  answer?: string;
  message?: string;
  retry_count?: number;
  /** type 为 interrupt 时携带待裁决内容。 */
  interrupt?: AgentInterrupt;
}

const TOKEN_KEY = "rag_token";

/**
 * 发起 SSE 请求，边读边回调。
 *
 * 用法：
 *   await streamAgentRun({
 *     question: "...",
 *     sessionId: "default",
 *     onEvent: (evt) => { ... },
 *     onDone: () => { ... },
 *     onError: (err) => { ... },
 *   });
 */
export async function streamAgentRun(params: {
  question: string;
  sessionId?: string;
  onEvent: (evt: SSEEvent) => void;
  onDone?: () => void;
  onError?: (err: Error) => void;
  signal?: AbortSignal;
}): Promise<void> {
  const token = localStorage.getItem(TOKEN_KEY);
  const resp = await fetch("/api/v1/agent/runs/stream", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
    body: JSON.stringify({
      question: params.question,
      session_id: params.sessionId ?? "default",
    }),
    signal: params.signal,
  });

  if (!resp.ok || !resp.body) {
    throw new Error(`SSE 请求失败: ${resp.status}`);
  }

  await consumeSSE(resp.body, params.onEvent);
  params.onDone?.();
}

/**
 * SSE resume 端点。
 *
 * 传 decision 表示这是在回答人工介入：accept 采纳当前草稿，rewrite 让
 * Generator 按 Reviewer 的反馈重写。不传则按 checkpoint 继续执行。
 */
export async function streamAgentResume(params: {
  runId: string;
  decision?: string;
  onEvent: (evt: SSEEvent) => void;
  onDone?: () => void;
  onError?: (err: Error) => void;
  signal?: AbortSignal;
}): Promise<void> {
  const token = localStorage.getItem(TOKEN_KEY);
  const resp = await fetch("/api/v1/agent/runs/stream/resume", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
    body: JSON.stringify({ run_id: params.runId, decision: params.decision }),
    signal: params.signal,
  });

  if (!resp.ok || !resp.body) {
    throw new Error(`SSE 请求失败: ${resp.status}`);
  }

  await consumeSSE(resp.body, params.onEvent);
  params.onDone?.();
}

/**
 * 消费 ReadableStream，按 SSE 协议 (`data: {...}\n\n`) 解析每行 JSON。
 */
async function consumeSSE(
  body: ReadableStream<Uint8Array>,
  onEvent: (evt: SSEEvent) => void
): Promise<void> {
  const reader = body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;

    buffer += decoder.decode(value, { stream: true });

    // 按 \n\n 分割 SSE 事件
    let idx: number;
    while ((idx = buffer.indexOf("\n\n")) !== -1) {
      const rawEvent = buffer.slice(0, idx);
      buffer = buffer.slice(idx + 2);

      // SSE 格式: "data: {json}\n"
      const line = rawEvent
        .split("\n")
        .filter((l) => l.startsWith("data:"))
        .map((l) => l.slice(5).trim())
        .join("");

      if (!line) continue;
      try {
        const evt = JSON.parse(line) as SSEEvent;
        onEvent(evt);
      } catch {
        // 忽略非 JSON 行
      }
    }
  }
}
